File: src/lib/analyzer.ts
export type Complexity = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9

// The four character classes a password can draw from, with the size of
// each class's alphabet (used to estimate the brute-force search space).
const CHAR_SETS = [
  { name: "lowercase", regex: /[a-z]/, size: 26 },
  { name: "uppercase", regex: /[A-Z]/, size: 26 },
  { name: "digit", regex: /[0-9]/, size: 10 },
  { name: "symbol", regex: /[^a-zA-Z0-9]/, size: 33 }
] as const


const clamp = (value: number, min: number, max: number) => Math.min(max, Math.max(min, value))

// Counts characters that extend a run of 3+ ascending, descending, or
// repeated code points (e.g. "abc", "321", "aaaa") — these are predictable.
const countSequential = (password: string) => {
  let count = 0
  for (let i = 2; i < password.length; i++) {
    const a = password.charCodeAt(i - 2)
    const b = password.charCodeAt(i - 1)
    const c = password.charCodeAt(i)
    const ascending = b - a === 1 && c - b === 1
    const descending = a - b === 1 && b - c === 1
    const repeated = a === b && b === c
    if (ascending || descending || repeated) count++
  }
  return count
}

const feedbackMap = {
  short: ["silly", "poor", "compromised", "fragile", "descent", "safe", "solid", "secure", "overkill", "paranoid"],
  time: [
    "no time",
    "split seconds",
    "less than 1 second",
    "few seconds",
    "less than minute",
    "less than an hour",
    "few hours",
    "a busy day",
    "a week",
    "a month",
    "few months",
    "about a year",
    "couple of years",
    "a decade",
    "a century",
    "thousands of years",
    "half a million years",
    "a million years",
    "100 million years",
    "a billion years",
    "a trillion years",
    "as long as sun shines",
    "as long as galaxy expands",
    "forever"
  ]
}

/**
 * Estimates password strength on a 0–9 scale.
 *
 * Combines three signals:
 *  - search-space entropy: length × log2(size of the combined alphabet),
 *    which rewards both longer passwords and more character classes;
 *  - character variety: the ratio of unique characters, penalizing
 *    passwords padded with repeats;
 *  - predictability: a penalty for ascending/descending/repeated runs.
 */
const analyzePassword = (password: string): Complexity => {
  if (!password) return 0

  // How many of the 4 classes are present and the combined alphabet size.
  const present = CHAR_SETS.filter((set) => set.regex.test(password))
  const poolSize = present.reduce((sum, set) => sum + set.size, 0)
  if (poolSize === 0) return 0

  const entropyBits = password.length * Math.log2(poolSize)

  // 0..1: lower when the same characters are reused.
  const uniqueRatio = new Set(password).size / password.length

  // 0..0.5: higher when more of the password sits in predictable runs.
  const sequencePenalty = clamp(countSequential(password) / password.length, 0, 0.5)

  // Entropy discounted for the structural weaknesses above.
  const effectiveBits = entropyBits * uniqueRatio * (1 - sequencePenalty)

  // Map effective bits onto 0–9. ~90 bits is treated as "max strength".
  const score = clamp(Math.round((effectiveBits / 90) * 9), 0, 9)

  return score as Complexity
}

export type Feedback = {
  estimate: string
  timeframe: string
}

/**
 * Turns a strength score into human-readable feedback: a one-word verdict and
 * a rough "time to crack" estimate.
 *
 *  - `estimate` is indexed directly by the 0–9 score, so it tracks complexity.
 *  - `timeframe` is driven by `length` but scaled by a coefficient so it doesn't
 *    read as a flat, hardcoded line. The coefficient mirrors the idea of
 *    "1.0 for a single charset, +0.25 per extra charset, ±0.25 for entropy":
 *    those signals are already folded into the score, so we expand the score
 *    back out across the same ~0.75 (weak) → ~2.0 (rich) range. Richer
 *    passwords stretch each character into a longer estimate; weaker ones
 *    compress it.
 */
const generateFeedback = (length: number, score: Complexity): Feedback => {
  const estimate = feedbackMap.short[clamp(score, 0, feedbackMap.short.length - 1)]

  const coefficient = 0.75 + (score / 9) * 1.25
  const timeIndex = clamp(Math.round(length * coefficient), 0, feedbackMap.time.length - 1)
  const timeframe = feedbackMap.time[timeIndex]

  return { estimate, timeframe }
}

export { generateFeedback }
export default analyzePassword

File: packages/cli-go/internal/password/analyzer.go
package password

import (
	_ "embed"
	"encoding/json"
	"math"
)

// feedback.json is copied from packages/core/feedback.json at build time
// (see Makefile `sync`). core is the single source of truth for the strings;
// this is a generated, gitignored copy so `go:embed` can reach it.
//
//go:embed feedback.json
var feedbackJSON []byte

type feedbackStrings struct {
	Short []string `json:"short"`
	Time  []string `json:"time"`
}

var feedback = loadFeedback()

func loadFeedback() feedbackStrings {
	var fb feedbackStrings
	_ = json.Unmarshal(feedbackJSON, &fb)
	return fb
}

type Complexity int

type charClass struct {
	size int
	test func(rune) bool
}

func isLower(r rune) bool { return r >= 'a' && r <= 'z' }
func isUpper(r rune) bool { return r >= 'A' && r <= 'Z' }
func isDigit(r rune) bool { return r >= '0' && r <= '9' }

// The four character classes, mirroring CHAR_SETS in packages/core/analyzer.ts.
// "symbol" is anything that is not a letter or digit ([^a-zA-Z0-9]).
var charClasses = []charClass{
	{size: 26, test: isLower},
	{size: 26, test: isUpper},
	{size: 10, test: isDigit},
	{size: 33, test: func(r rune) bool { return !isLower(r) && !isUpper(r) && !isDigit(r) }},
}

func clampFloat(value, min, max float64) float64 {
	return math.Min(max, math.Max(min, value))
}

func clampInt(value, min, max int) int {
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

// countSequential counts characters extending a run of 3+ ascending,
// descending, or repeated code points (e.g. "abc", "321", "aaaa").
func countSequential(runes []rune) int {
	count := 0
	for i := 2; i < len(runes); i++ {
		a, b, c := runes[i-2], runes[i-1], runes[i]
		ascending := b-a == 1 && c-b == 1
		descending := a-b == 1 && b-c == 1
		repeated := a == b && b == c
		if ascending || descending || repeated {
			count++
		}
	}
	return count
}

// Analyze estimates password strength on a 0–9 scale. Port of analyzePassword
// in packages/core/analyzer.ts; the math is kept identical.
func Analyze(pw string) Complexity {
	if pw == "" {
		return 0
	}

	runes := []rune(pw)

	poolSize := 0
	for _, class := range charClasses {
		for _, r := range runes {
			if class.test(r) {
				poolSize += class.size
				break
			}
		}
	}
	if poolSize == 0 {
		return 0
	}

	length := float64(len(runes))
	entropyBits := length * math.Log2(float64(poolSize))

	unique := make(map[rune]struct{})
	for _, r := range runes {
		unique[r] = struct{}{}
	}
	uniqueRatio := float64(len(unique)) / length

	sequencePenalty := clampFloat(float64(countSequential(runes))/length, 0, 0.5)

	effectiveBits := entropyBits * uniqueRatio * (1 - sequencePenalty)

	score := clampFloat(math.Round((effectiveBits/90)*9), 0, 9)
	return Complexity(int(score))
}

type Feedback struct {
	Estimate  string
	Timeframe string
}

// GenerateFeedback turns a score into a one-word verdict and a time-to-crack
// estimate. Port of generateFeedback in packages/core/analyzer.ts.
func GenerateFeedback(length int, score Complexity) Feedback {
	s := int(score)

	estimate := feedback.Short[clampInt(s, 0, len(feedback.Short)-1)]

	coefficient := 0.75 + (float64(s)/9)*1.25
	timeIndex := clampInt(int(math.Round(float64(length)*coefficient)), 0, len(feedback.Time)-1)
	timeframe := feedback.Time[timeIndex]

	return Feedback{Estimate: estimate, Timeframe: timeframe}
}
